src/main/java/com/eerussianguy/blazemap/engine/cache/ChunkMDCache.java

```diff
@@ -13,49 +15,142 @@
         BlazeRegistry.Key<DataType> key = UnsafeGenerics.stripKey(datum.getID());
-        MasterDatum old = data.get(key);
-        if(!datum.equals(old)) {
-            data.put(key, datum);
-            setDirty();
-            return true;
+
+        wLock.lock();
+        try {
+            MasterDatum old = data.get(key);
+
+            if(!datum.equals(old)) {
+                setDirty(true);
+                data.put(key, datum);
+
+                return true;
+            }
+            return false;
+
+        } finally {
+            wLock.unlock();
         }
-        return false;
     }
 
-    protected void setDirty() {}
-
     public boolean isEmpty() {
-        return data.isEmpty();
+        rLock.lock();
+        try {
+            return data.isEmpty();
+
+        } finally {
+            rLock.unlock();
+        }
     }
 
     public Set<BlazeRegistry.Key<DataType>> keys() {
-        return data.keySet();
+        rLock.lock();
+        try {
+            return data.keySet();
+
+        } finally {
+            rLock.unlock();
+        }
     }
 
     public List<MasterDatum> data() {
-        return data.values().stream().toList();
+        rLock.lock();
+        try {
+            return data.values().stream().toList();
+
+        } finally {
+            rLock.unlock();
+        }
     }
 
     public ChunkMDCache clear() {
-        data.clear();
-        return this;
+        wLock.lock();
+        try {
+            setDirty(true);
+            data.clear();
+
+            return this;
+
+        } finally {
+            wLock.unlock();
+        }
     }
 
     public ChunkMDCache copy() {
         return copyInto(new ChunkMDCache());
     }
 
     public ChunkMDCache copyInto(ChunkMDCache clone) {
-        clone.data.clear();
-        clone.data.putAll(data);
-        return clone;
+        rLock.lock();
+        clone.wLock.lock();
+
+        try {
+            clone.setDirty(true);
+
+            clone.data.clear();
+            clone.data.putAll(data);
+            return clone;
+
+        } finally {
+            clone.wLock.unlock();
+            rLock.unlock();
+        }
+    }
+
+    /**
+     * Mark this chunk as either "dirty" or "not dirty" (ie "clean").
+     *
+     * A chunk can only be marked dirty if already holding the write lock, however can be marked as
+     * clean from either the read or write lock (eg if cache has been read out to file to save).
+     *
+     * Note: As isDirty() is _not_ protected by the cache lock (though this.dirty is marked volatile to
+     * synchronise its value between threads), placement of calls to setDirty() need to be considered in the
+     * context of providing the most useful value to other threads checking this chunks current cleanliness status
+     * (knowing they won't then be able to act further on this chunk without first acquiring its lock)
+     */
+    protected void setDirty(boolean dirty) {
+        rLock.lock();
```

Comment:
Read Lock is unnecessary because we already enforce Write Lock is being held, right?
So, if we're already holding the write lock, it's impossible for anyone to lock for Read until we let go of Write.

Reply:
So this is why I put in that big comment at the top there, because this is _not_ that normal scenario. 

Note specifically (emphasis added):
>  A chunk can only be marked dirty if already holding the write lock, however **can be marked as clean from either the read or write lock** (eg if cache has been read out to file to save).

In other words, this function is being allowed to be called from a read lock _in addition to_ a write lock, as long as the read lock is _only_ setting the `dirty` state to `false` (ie `clean`). This is so that, once the cache has been read out to the stream that's writing to disk (under a read lock for the cache), the `dirty` status can be updated to reflect this _without_ having to unnecessarily swap to a write lock.

The check below makes sure it's only ever allowed to be set `clean` from a non-write lock, given no actual changes to the cache contents are needed for that to be the case, and will throw an exception calling you out if you tried to dirty the cache without holding the write lock first.

Reply:
I'm only grabbing the lock again to make sure that there is definitely _any_ lock being held at all and the dev didn't forget to lock completely. We don't want threads to be setting the variable to different values at the same time! 

And even though it's allowed to be written to under only a read lock, there's only _one_ value that it can be changed to under a read lock, so the worst that can happen is multiple threads simultaneously setting `dirty` to `false`. But if trying to change to `true`, a write lock must already be held from the calling function, and a write lock is an exclusive lock so there can be no contention from different threads trying to change the variable to a different value at the same time.

Reply:
Ah, very well.